src/Microsoft.ML/MemoryCollection.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.EntryPoints;
+
+namespace Microsoft.ML
+{
+    public class MemoryCollection<TInput> : ILearningPipelineLoader
+        where TInput : class
+    {
+        public ILearningPipelineStep ApplyStep(ILearningPipelineStep previousStep, Experiment experiment)
+        {
+            Contracts.Assert(previousStep == null);
+            _dataViewEntryPoint = new Data.DataViewReference();
+            var importOutput = experiment.Add(_dataViewEntryPoint);
+            return new MemoryCollectionPipelineStep(importOutput.Data);
+        }
+
+        private readonly IList<TInput> _listCollection;
+        private readonly IEnumerable<TInput> _enumerableCollection;
+
+        private Data.DataViewReference _dataViewEntryPoint;
+        private IDataView _dataView;
+
+        public MemoryCollection(IList<TInput> collection)
+        {
+            //need validation at some point
```

Comment:
>//need validation at some point [](start = 12, length = 31)

How about we at least validate to ensure that these aren't `null`, with a `Contracts.CheckParamValue`? Here and in the other constructor.